sympy/physics/continuum_mechanics/beam.py

```diff
@@ -0,0 +1,567 @@
+    >>> b.deflection()
+    (-7*x + SingularityFunction(x, 0, 3)/2 - SingularityFunction(x, 2, 4)/4 + 3*SingularityFunction(x, 4, 3)/2)/(E*I)
+    """
+
+    def __init__(self, length, elastic_modulus, second_moment):
+        self._length = length
+        self._elastic_modulus = elastic_modulus
+        self._second_moment = second_moment
+        self._boundary_conditions = {'deflection': [], 'moment': [], 'slope': []}
+        self._load = 0
+
+    def __str__(self):
+        str_sol = 'Beam(%s, %s, %s)' % (sstr(self._length), sstr(self._elastic_modulus), sstr(self._second_moment))
+        return str_sol
+
+    __repr__ = __str__
+
+    @property
+    def length(self):
+        """Length of the Beam."""
+        return self._length
+
+    @length.setter
+    def length(self, l):
+        self._length = sympify(l)
+
+    @property
+    def elastic_modulus(self):
+        """Young's Modulus of the Beam. """
+        return self._elastic_modulus
+
+    @elastic_modulus.setter
+    def elastic_modulus(self, e):
+        self._elastic_modulus = sympify(e)
+
+    @property
+    def second_moment(self):
+        """Second moment of area of the Beam. """
+        return self._second_moment
+
+    @second_moment.setter
+    def second_moment(self, i):
+        self._second_moment = sympify(i)
+
+    @property
+    def boundary_conditions(self):
+        """
+        Returns a dictionary of boundary conditions applied on the beam.
+        """
+        return self._boundary_conditions
+
+    def apply_boundary_conditions(self, moment = [], slope = [], deflection = []):
+        """
+        Takes the boundary conditions of the beam bending problem as input.
+        The boundary conditions should be passed as keyworded arguments.
+        It is suggested to use ``moment``, ``slope`` and ``deflection`` as the
+        keywords while providing the boundary conditions of the moment curve,
+        slope curve and the deflection curve respectively as inputs.
+        A tuple in the form of (location, value) can be used to pass the
+        location and the value of a boundary condition. A list of such tuples
+        is passed under each keywords.
+
+
+        Examples
+        ========
+        >>> from sympy.physics.continuum_mechanics.beam import Beam
+        >>> from sympy import Symbol
+        >>> E = Symbol('E')
+        >>> I = Symbol('I')
+        >>> b = Beam(4, E, I)
+        >>> b.apply_boundary_conditions(moment = [(0, 4), (4, 0)], deflection = [(0, 2)], slope = [(0, 1)])
+
+        Here ``moment = [(0, 4), (4, 0)]`` means the moment due to the
+        applied loads at point ``0`` and ``4`` are `4` and `0` respectively.
+        Similarly, ``deflection = [(0, 2)]`` means the deflection of the beam
+        at point ``0`` is ``2``.
+
+        >>> b.boundary_conditions
+        {'deflection': [(0, 2)], 'moment': [(0, 4), (4, 0)], 'slope': [(0, 1)]}
+        >>> b.boundary_conditions['moment']
+        [(0, 4), (4, 0)]
```

Comment:
I think the word `apply` implies that you will add (append) boundary conditions to existing ones. The word `set` would better describe overwriting existing boundary conditions.

I'd like you to think carefully about not using set/apply type methods though. For example, you have a data structure for the boundary conditions that is a dictionary of lists of tuples.

`b = {'deflection': [(0, 2)], 'moment': [(0, 4), (4, 0), (4, 3), (5, 0)], 'slope': [(0, 1)]}`

A user can do all the necessary operations like so:

overwrite a set of boundary conditions

`b['deflection'] = [(0, 1), (3, 4)]`

append/apply new boundary conditions

`b['slope'].append((5, 6)]`

remove boundary conditions of one type:

`b['slope'] = []`

remove all boundary conditions

`b = {}`

overwrite a specific bc:

`b['slope'][0] = (4, 5)`

etc

If you write this out in the documentation of the class, then no methods are needed to update this data structure.

I'm not completely convinced that you have the best data structure though (at least for easy of use) but I don't see why you need to create all of these extra methods for setting/applying things in this case.

Some other ideas:

```python
b.deflection_bc = [{'loc': 1, 'val`: 2}, {'loc': 2, 'val': 3}]
b.deflection_bc = [(1, 2), (2, 3)]
b.fix('deflection', 1, 2)
b.fix_deflection(1, 2)
```

A list of tuples is pretty succint and simple. The mutability of the list is nice so you can append and remove the bc's. The tuple  will be difficult to remember whether the location or value comes first.

Another idea:

```python
DeflectionBC = namedtuple(loc=0, val=0)
bc = DeflectionBC(1, 2)
bc.loc
bc.val
beam.boundary_conditions.append(bc)
beam.boundar_conditions.append(Deflection(loc=3, val=4))
```